iceoryx_utils/include/iceoryx_utils/posix_wrapper/posix_call.hpp

```diff
@@ -0,0 +1,198 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_UTILS_POSIX_WRAPPER_POSIX_CALL_HPP
+#define IOX_UTILS_POSIX_WRAPPER_POSIX_CALL_HPP
+
+#include "iceoryx_utils/cxx/attributes.hpp"
+#include "iceoryx_utils/cxx/expected.hpp"
+#include "iceoryx_utils/cxx/string.hpp"
+
+#include <cstdint>
+#include <cstring>
+
+namespace iox
+{
+namespace posix
+{
+template <typename ReturnType, typename... FunctionArguments>
+class PosixCallBuilder;
+
+/// @brief result of a posix call
+template <typename T>
+struct PosixCallResult
+{
+    static constexpr uint32_t ERROR_STRING_SIZE = 128u;
+    static const PosixCallResult INVALID_STATE;
+
+    PosixCallResult() noexcept = default;
+
+    /// @brief returns the result of std::strerror(errnum) which acquires a
+    ///        human readable error string
+    cxx::string<ERROR_STRING_SIZE> getHumanReadableErrnum() const noexcept;
+
+    /// @brief the return value of the posix function call
+    T value;
+
+    /// @brief the errno value which was set by the posix function call
+    int32_t errnum = -1;
+};
+
+template <typename T>
+PosixCallResult<T> const PosixCallResult<T>::INVALID_STATE{{}, -1};
+
+namespace internal
+{
+template <typename ReturnType, typename... FunctionArguments>
+PosixCallBuilder<ReturnType, FunctionArguments...> createPosixCallBuilder(ReturnType (*posixCall)(FunctionArguments...),
+                                                                          const char* posixFunctionName,
+                                                                          const char* file,
+                                                                          const int32_t line,
+                                                                          const char* callingFunction) noexcept;
+
+template <typename ReturnType>
+struct PosixCallDetails
+{
+    const char* posixFunctionName = nullptr;
+    const char* file = nullptr;
+    int32_t line = 0;
+    const char* callingFunction = nullptr;
+    bool hasSuccess = true;
+
+    PosixCallResult<ReturnType> result;
+};
+} // namespace internal
+
+/// @brief Calling a posix function with automated error handling. If the posix function returns
+///        void you do not need to use posixCall since it cannot fail, (see: man errno).
+///        We use a builder pattern to create a design which sets the usage contract so that it
+///        cannot be used in the wrong way.
+/// @code
+///        iox::posix::posixCall(sem_timedwait)
+///             .call(handle, timeout)
```

Comment:
I'm not quite sure with just `call()` when there are arguments. It might be a bit unintuitive. I'm just thinking it there is a another name which suits both cases well (`doIt(...)`, `makeItSo(...)`?) or if there should be a second method.

Reply:
We also could use the call operator, then it would look like
```cpp
iox::posix::posixCall(sem_timedwait)(handle, timeout)
```

Reply:
hahaha, when you can't come up with a good name just remove the name. I also like it.